src/bsoncxx/types/bson_value/value.hpp

```diff
@@ -116,11 +116,21 @@ class BSONCXX_API value {
     ///
     value(bsoncxx::document::view v);
 
+    ///
+    /// Constructs a BSON document value.
+    ///
+    value(bsoncxx::document::value v);
```

Comment:
Take this example
```
builder::list::document b{"foo", builder::list::array{"bar"}};
```
`builder::list::array` needs to be implicitly converted to a `builder::list::document`. This will call the constructor
```
// document.hpp
template <typename T>
document(T v) : _value{v} {}
```
Here, `_value` is a `bson_value::value`. `array` can be implicitly converted to a `bson_value::value` but it can't be implicitly converted _again_ to a `bson_value::view`. That's the heart of the problem. The reverse is true, as well.

One solution I considered was to create specialized constructors. That is
```
// document.hpp
#include "array.hpp"
document(builder::list::array arr) : _value{arr.extract()} {}

// array.hpp
#include "document.hpp"
array(builder::list::document doc) : _value{doc.extract()} {}
```
But this causes a circular dependency between the builders. So far, this is the best solution I've come across.

Reply:
That makes sense!  Thanks for explaining, seems fine to me.